[orderbook][top-bids] fixing top bids processing

ochaloup opened the pull request:
It seems that the top-bids are wrongly calculated when updated by watchers.

Based on the comments and code
https://github.com/blockworks-foundation/mango-explorer/blob/v3.2.8/mango/orders.py#L238
the top-bid should be at index `0`. That's to reverse the order of the list. That's what the `init` function in `OrderBook` does. That's what is not done when updated in watchers
https://github.com/blockworks-foundation/mango-explorer/blob/v3.2.8/mango/watchers.py#L224
the order reverse is not processed and the top-bid then returns the bid at tail.

Not sure if `reversed` is correctly used at pandas dataframe creation - https://github.com/blockworks-foundation/mango-explorer/blob/v3.2.8/mango/orders.py#L284

OpinionatedGeek commented:
Thanks!

As background, the `OrderBook` is newer than the `Watcher` code and it looks like I missed this when updating things. Apologies, and thanks for finding and fixing it.

I was going to just immediately apply this PR but then I thought I'd pick your brains on a couple of points.

1. Would it be better to have getter/setter properties hiding a private member for `bids` and `asks`? The logic you have in `update_bids()`, say, would then be in the `bids` setter. It might be a more obvious interface and it might help stop me making this mistake again.
2. Can you think of any useful tests around this? I'm trying to add more tests as I go along now, and I realise the `OrderBook` currently has none.

I'm happy to apply this PR as-is (thanks again), but I figured I'd ask those questions first.

ochaloup commented:
@OpinionatedGeek +1, I will elaborate on the PR based on your suggestions. Hopefully today/tomorrow I could have the updated code.

ochaloup commented:
@OpinionatedGeek I started with this but I found that I doubt about one thing - implicit ordering by `order.id` - the `init` considers the ordering by price being implicit to order id
https://github.com/blockworks-foundation/mango-explorer/blob/v3.2.8/mango/orders.py#L234
is it so? Is it something the Mango Explorer count with?

OpinionatedGeek commented:
Hi,

That's intentional. It would benefit from testing, of course, but it's intentional.

The reasoning is interesting, and worth knowing. Mango follows the Serum practice for order IDs, and they're a kind of compound piece of data.

The ID is a 16-byte 'int' where:
* The low 8 bytes encodes the sequence number from the market (just an auto-increment int, I think).
* The high 8 bytes encodes the price.

The result of this is an int that can be used for sorting orders, and will sort them by price, and then sort entries with the same price by 'time priority' so that earlier orders are filled before later orders.

You can see an example of parsing the 'compound ID' here: https://github.com/blockworks-foundation/mango-explorer/blob/d49294f2cd8d3521062e228013e240b193909ca8/mango/layouts/layouts.py#L237

ochaloup commented:
@OpinionatedGeek thanks a lot for elaboration on how the things work under the hood. I changed the tests to follow this detail and I hope the PR is ready for your review now.
(I searched the code and haven't found a method to be used for the fake construction of order id so I added one to `fakes.py`)

OpinionatedGeek commented:
Thank you so much for doing this!

ochaloup commented:
@OpinionatedGeek no problem. Thanks for the mango explorer (+thanks for fixing the linting issues, my bad on not fixing them).
